leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/impl/QueuedRequestSender.java

```diff
@@ -0,0 +1,282 @@
+/*******************************************************************************
+ * Copyright (c) 2015 Bosch Software Innovations GmbH and others.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * and Eclipse Distribution License v1.0 which accompany this distribution.
+ *
+ * The Eclipse Public License is available at
+ *    http://www.eclipse.org/legal/epl-v10.html
+ * and the Eclipse Distribution License is available at
+ *    http://www.eclipse.org/org/documents/edl-v10.html.
+ *
+ * Contributors:
+ *     Alexander Ellwein, Daniel Maier, Balasubramanian Azhagappan (Bosch Software Innovations GmbH)
+ *                                - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.leshan.server.queue.impl;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+
+import org.eclipse.leshan.core.node.LwM2mNode;
+import org.eclipse.leshan.core.observation.Observation;
+import org.eclipse.leshan.core.request.BindingMode;
+import org.eclipse.leshan.core.request.DownlinkRequest;
+import org.eclipse.leshan.core.response.ErrorCallback;
+import org.eclipse.leshan.core.response.LwM2mResponse;
+import org.eclipse.leshan.core.response.ResponseCallback;
+import org.eclipse.leshan.server.Stoppable;
+import org.eclipse.leshan.server.client.Client;
+import org.eclipse.leshan.server.client.ClientRegistry;
+import org.eclipse.leshan.server.client.ClientRegistryListener;
+import org.eclipse.leshan.server.client.ClientUpdate;
+import org.eclipse.leshan.server.observation.ObservationRegistry;
+import org.eclipse.leshan.server.observation.ObservationRegistryListener;
+import org.eclipse.leshan.server.queue.MessageStore;
+import org.eclipse.leshan.server.queue.QueuedRequest;
+import org.eclipse.leshan.server.queue.QueuedRequestFactory;
+import org.eclipse.leshan.server.request.LwM2mRequestSender;
+import org.eclipse.leshan.util.Validate;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This sender is a special implementation of a {@link LwM2mRequestSender}, which is aware of a LwM2M client's binding
+ * mode. If the client supports "Q" (queue) mode, then the request is processed using the internal queue, i.e. the
+ * request is enqueued first and then sent if a client is back online. If the client does not support the queue mode in
+ * its binding, then the delegate sender is called in order to send the request.
+ */
+public class QueuedRequestSender implements LwM2mRequestSender, Stoppable {
+    private static final Logger LOG = LoggerFactory.getLogger(QueuedRequestSender.class);
```

Comment:
Yes you are correct. As queue mode is a mandatory feature that the server should support, it would be great to have the queue mode functionality as part of the standard `LwM2mRequestSender` and use the AsyncRequestObserver . Thought of doing one step at a time. As of now left this Sender as an extension to `LwM2mRequestSender`. It can be done as the next PR once this one is accepted. Also as discussed by @sophokles73 and @jvermillard in https://github.com/eclipse/leshan/pull/47#issuecomment-157369748, it would be great to have API change in `LwM2mRequestSender`.